### PR TITLE
Refactor stateHandler to preserve integrity

### PR DESCRIPTION
**Pull request by epinxteren**
Create KeyVaueStore with unit test
Remove many setter in the stateHandler.

### CODE REVIEW

File: src/Service/StateHandler.php [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    const RESPONSE_ERROR_MESSAGE = 'error_response_message';
+
+    private $store;
+
+    public function __construct(ValueStore $store)
+    {
+        $this->store = $store;
+    }
+
+    public function saveRegistrationRequest(Request $request, ReceivedAuthnRequest $authnRequest, $stepupRequestId)
```
Comment (MKodde) [inline src/Service/StateHandler.php]: Wasn't one of the points of implementing the `saveRegistrationRequest` and `saveAuthenticationRequest` in the State handler to be able to protect the integrity in a more elegant way? 

I expected to see more input validation on the input of these methods. Maybe this is not needed as their state is validated elsewhere. Just being curious here.

File: src/Controller/SSOController.php [hunk 2]
```diff
@@ -104,13 +103,11 @@ public function ssoAction(Request $request)
...
     private function authenticationAction(Request $request, ReceivedAuthnRequest $originalRequest)
     {
-        $this->stateHandler
-            ->setRequestId($originalRequest->getRequestId())
-            ->setStepupRequestId($this->getOrGenerateStepupRequestId($request))
-            ->setRequestServiceProvider($originalRequest->getServiceProvider())
-            ->setRelayState($request->get(AuthnRequest::PARAMETER_RELAY_STATE, ''))
-            ->setRequestTypeAuthentication($originalRequest->getNameId())
-        ;
+        $this->stateHandler->saveAuthenticationRequest(
```
Comment (MKodde) [inline src/Controller/SSOController.php:107]: Definitely cleaner, more readable than the previous solution.

File: src/Exception/RuntimeException.php [hunk 3]
```diff
@@ -26,12 +26,11 @@ public static function shouldNotRegister()
         return new self('The current application context does not require a registration');
     }
 
-    public static function requestTypeAlreadyGiven($previous, $type)
+    public static function requestTypeAlreadyKnown($previous)
```
Comment (MKodde) [inline src/Exception/RuntimeException.php:29]: Why did you change this? Was the error to specific previously or are you no longer able to get to the newly set type?

File: src/Service/StateHandlerInterface.php [hunk 4]
```diff
@@ -15,71 +15,74 @@
...
+    public function getRelayState();
 
     /**
-     * Set the current request type as registration flow.
-     *
-     * @return $this
+     * @return string
      */
-    public function setRequestTypeRegistration();
+    public function getIdentityNameId();
```
Comment (MKodde) [inline src/Service/StateHandlerInterface.php]: In my opinion, get/set SubjectNameId was a better (closer to SAML spec) name for these methods.

**Review by MKodde — COMMENTED** [review]
Some comments below. We can also discuss this IRL if you prefer :)

File: src/Exception/RuntimeException.php (hunk 3 above)
Comment (epinxteren) [reply src/Exception/RuntimeException.php:29]: Yea it was to specific, also wanted to protect against setting the same request type again.

File: src/Service/StateHandler.php (hunk 1 above)
Comment (epinxteren) [reply src/Service/StateHandler.php]: Good question. 

1. The controller validates input data like the AuthnRequest. 
2. This StateHandler protects the integrity of the current state and when changes can happen to it. Not if if the given variables are correct. I did add some minor assertions to know that the given variables are of the correct scaler type.

It's was not ideal that the Request is passed to this function, what can be considered as insecure user data, wat of course always should be validated. To fix this issue i moved the relay state property to the controller so it's given as parameter instead. 

But i don't think the controller can validate the relay state. I added some documentation why not, and what assumptions i have taken here.

File: src/Service/StateHandlerInterface.php (hunk 4 above)
Comment (epinxteren) [reply src/Service/StateHandlerInterface.php]: 👍  will change it

**Review by MKodde — APPROVED** [review]
